### PR TITLE
Move Vagrant provider to Flannel

### PR DESCRIPTION
**Pull request by derekwaynecarr**
This PR makes the following changes:
- installs etcd on host to hold flannel config
- install configure flannel on master and nodes
- update salt to take a flag to not remove host etcd setup

Fixes https://github.com/kubernetes/kubernetes/issues/13920
Fixes https://github.com/kubernetes/kubernetes/issues/11119

Once this lands, I will back port it to 1.0.x release stream.

### CODE REVIEW

**Comment by derekwaynecarr** [comment]
As for why I cannot use the kubelet provisioned etcd to hold flannel configuration, its a chicken/egg problem since I need to configure flannel before I install/run docker and/or kubelet.

**Comment by bprashanth** [comment]
Fyi https://github.com/kubernetes/kubernetes/pull/13877
Not that it's going in soon, but sometime we will solve the chicked and egg problem for you. in fact flannel won't need access to etcd.

**Comment by derekwaynecarr** [comment]
@bprashanth - I would like this to go in before #13877 because it's easier to backport to 1.0.x stream, works, and unblocks folks.  I hope this change will make it easier to then have vagrant included in the changes for #13877

File: cluster/saltbase/salt/etcd/init.sls [hunk 1]
```diff
@@ -34,6 +26,28 @@ delete_etcd_initd:
...
+    - name: etcd
+    - enable: None
+
+{% endif %}
+
+touch /var/log/etcd.log:
+  cmd.run:
+    - creates: /var/log/etcd.log
+
+/var/etcd:
```
Comment (roberthbailey) [inline cluster/saltbase/salt/etcd/init.sls]: Should this be in an else block? For the platforms that don't have a host etcd, we don't need this rule to execute.

File: cluster/vagrant/provision-network-minion.sh [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+# provision-network-minion configures flannel on the minion
+function provision-network-minion {
+
+  echo "Provisioning network on minion"
```
Comment (roberthbailey) [inline cluster/vagrant/provision-network-minion.sh]: All of these steps run on both the master and minion, right? It seems like there's quite a bit of duplicated shell script as a result, which might be better off combined into a single function.

File: cluster/saltbase/salt/etcd/init.sls (hunk 1 above)
Comment (derekwaynecarr) [inline cluster/saltbase/salt/etcd/init.sls]: No.

Even though we run the default kubernetes etcd via a pod, we bind mount that directory in for logging, so we need to touch the files first in order to allow the mount to proceed.

See https://github.com/kubernetes/kubernetes/blob/master/cluster/saltbase/salt/etcd/etcd.manifest#L41

**Comment by tsl0922** [comment]
@derekwaynecarr I can confirm this pr fixes #11119 for me, everything works fine, except that there is a failed unit(`docker-storage-setup.service`) on both master and minion, log:

```
[root@kubernetes-minion-1 vagrant]# journalctl -ru docker-storage-setup.service
-- Logs begin at Tue 2015-02-24 15:27:58 UTC, end at Wed 2015-09-16 00:04:44 UTC. --
Sep 15 23:25:00 kubernetes-minion-1 systemd[1]: docker-storage-setup.service failed.
Sep 15 23:25:00 kubernetes-minion-1 systemd[1]: Unit docker-storage-setup.service entered failed state.
Sep 15 23:25:00 kubernetes-minion-1 systemd[1]: Failed to start Docker Storage Setup.
Sep 15 23:25:00 kubernetes-minion-1 systemd[1]: docker-storage-setup.service: main process exited, code=exited, status=1/FAILURE
Sep 15 23:25:00 kubernetes-minion-1 docker-storage-setup[3186]: There is not enough free space in volume group vg_vagrant to create data volume of size MIN_DATA_SIZE=2G.
Sep 15 23:25:00 kubernetes-minion-1 docker-storage-setup[3186]: Logical volume "docker-poolmeta" created
Sep 15 23:24:59 kubernetes-minion-1 docker-storage-setup[3186]: Rounding up size to full physical extent 64.00 MiB
Sep 15 23:24:59 kubernetes-minion-1 systemd[1]: Starting Docker Storage Setup...
```

**Comment by derekwaynecarr** [comment]
@tsl0922 - the docker storage setup unit failure is fine, and results in using devicemapper.  I will move to overlayfs when its stable.  Thanks for testing the feature!

File: cluster/saltbase/salt/etcd/init.sls (hunk 1 above)
Comment (roberthbailey) [inline cluster/saltbase/salt/etcd/init.sls]: Got it, thanks.

**Comment by roberthbailey** [comment]
LGTM.

**Comment by nirving-versent** [comment]
Works for me, although still have issues accessing https://10.245.1.2/api/v1/proxy/namespaces/kube-system/services/kube-dns, but think it is related to something else.

**Pull request merged**